zebra-chain/src/primitives/zcash_history.rs

```diff
@@ -0,0 +1,194 @@
+    transaction::Transaction,
+};
+
+/// A MMR Tree using zcash_history::Tree.
+///
+/// Currently it should not be used as a long-term data structure because it
+/// may grow undefinitely.
+pub struct Tree {
+    network: Network,
+    tree: zcash_history::Tree,
+}
+
+/// An encoded tree Node.
+pub struct Node {
+    node: [u8; zcash_history::MAX_NODE_DATA_SIZE],
+}
+
+impl Node {
+    /// Convert a Node into a zcash_history::Entry.
+    fn to_entry(&self, branch_id: ConsensusBranchId) -> Result<zcash_history::Entry, io::Error> {
+        zcash_history::Entry::from_bytes(branch_id.into(), self.node)
+    }
+}
+
+impl Tree {
+    /// Create a MMR tree with the given length.
+    ///
+    /// The `peaks` are the peaks of the MMR tree to build and their position in the
+    /// array representation of the tree.
+    /// The `extra` are extra nodes that enable removing nodes from the tree, and their position.
+    ///
+    /// Note that the length is usually larger than the length of `peaks` and `extra`, since
+    /// you don't need to pass every node, just the peaks of the tree (plus extra).
+    fn new(
+        network: Network,
+        branch_id: ConsensusBranchId,
+        length: u32,
+        peaks: &HashMap<u32, &Node>,
+        extra: &HashMap<u32, &Node>,
+    ) -> Result<Self, io::Error> {
+        let mut peaks_vec = Vec::new();
+        for (idx, node) in peaks {
+            peaks_vec.push((*idx, node.to_entry(branch_id)?));
+        }
+        let mut extra_vec = Vec::new();
+        for (idx, node) in extra {
+            extra_vec.push((*idx, node.to_entry(branch_id)?));
+        }
+        let tree = zcash_history::Tree::new(length, peaks_vec, extra_vec);
+        Ok(Tree { network, tree })
+    }
+
+    /// Append a new block to the tree, as a new leaf.
+    ///
+    /// `sapling_root` is the root of the Sapling note commitment tree of the block.
+    ///
+    /// Returns a vector of nodes added to the tree (leaf + internal nodes).
+    fn append_leaf(&mut self, block: &Block, sapling_root: &sapling::tree::Root) -> Vec<Node> {
+        let node_data = convert_block_to_librustzcash_data(block, self.network, sapling_root);
+        // TODO: handle error
+        let appended = self.tree.append_leaf(node_data).unwrap();
+
+        let mut new_nodes = Vec::new();
+        for entry in appended {
+            let mut node = Node {
+                node: [0; zcash_history::MAX_NODE_DATA_SIZE],
+            };
+            self.tree
+                .resolve_link(entry)
+                .expect("entry was just generated so it must be valid")
+                .data()
+                .write(&mut &mut node.node[..])
+                .expect("buffer was created with enough capacity");
+            new_nodes.push(node);
+        }
+        new_nodes
+    }
+
+    /// Append multiple blocks to the tree.
+    fn append_leaf_iter<'a>(
+        &mut self,
+        vals: impl Iterator<Item = (&'a Block, &'a sapling::tree::Root)>,
+    ) {
```

Comment:
We should be able to skip lifetimes here:
- `Arc` can be trivially cloned
- the sapling tree root should be a `Copy` type

```suggestion
    fn append_leaf_iter(
        &mut self,
        vals: impl Iterator<Item = (Arc<Block>, sapling::tree::Root)>,
    ) {
```

Reply:
Done